internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	this.ResultItems = make([]OutputType, 0, len(this.WorkItems))
+	this.RemainingItems = this.WorkItems
+	this.RunTime = 0
+
+	// If we are not overwriting records, query for stored records.
+	if !this.OverwriteRecords && this.RetrieveFunc != nil {
+		this.ResultItems, this.RemainingItems, err = this.RetrieveFunc(this.WorkItems)
+		if err != nil {
+			return JobOutput[InputType, OutputType]{}, err
+		}
+	}
+
+	if this.done == nil {
+		this.done = make(chan struct{})
+		this.Done = this.done
+	}
+
+	if this.WaitForCompletion {
+		err = this.run()
+		if err != nil {
+			return JobOutput[InputType, OutputType]{}, err
+		}
+
+		// If the context was canceled during execution, return immediately.
+		if this.Context.Err() != nil {
+			return JobOutput[InputType, OutputType]{}, nil
+		}
+
+		select {
+		case <-this.done:
+		default:
+			close(this.done)
+		}
+	} else {
+		go func() {
+			err = this.run()
+			if err != nil {
+				log.Error("Failure while running asynchronous job.")
+			}
+
+			select {
+			case <-this.done:
+			default:
+				close(this.done)
+			}
+		}()
+	}
+
+	return this.JobOutput, nil
+}
+
+func (this *Job[InputType, OutputType]) run() error {
+	err := this.Validate()
+	if err != nil {
+		return fmt.Errorf("Failed to validate job: '%v'.", err)
+	}
+
+	if len(this.RemainingItems) == 0 {
+		return nil
+	}
+
+	noLockWait := true
+	if this.LockKey != "" {
+		noLockWait = lockmanager.Lock(this.LockKey)
+		defer lockmanager.Unlock(this.LockKey)
+	}
+
+	// The context has been canceled while waiting for a lock, abandon this job.
+	if this.Context.Err() != nil {
+		return nil
+	}
+
+	// If we had to wait for the lock, then check again for stored records.
+	if !noLockWait && this.RetrieveFunc != nil {
+		var partialResults []OutputType = nil
+		partialResults, this.RemainingItems, err = this.RetrieveFunc(this.RemainingItems)
+		if err != nil {
+			return fmt.Errorf("Failed to re-check record storage before run: '%w'.", err)
+		}
+
+		// Collect the partial records from storage.
+		this.ResultItems = append(this.ResultItems, partialResults...)
+	}
+
+	if len(this.RemainingItems) == 0 {
+		return nil
+	}
+
+	// If we are overwriting records, then remove all the old records.
+	if this.OverwriteRecords && this.RemoveStorageFunc != nil {
+		processedItems := getProcessedItems(this.WorkItems, this.RemainingItems)
+		err = this.RemoveStorageFunc(processedItems)
+		if err != nil {
+			return fmt.Errorf("Failed to remove old job cache entries: '%w'.", err)
+		}
+	}
+
+	type PoolResult struct {
+		Input   InputType
+		Result  OutputType
+		RunTime int64
+		Error   error
+	}
+
+	poolResults, _, err := util.RunParallelPoolMap(this.PoolSize, this.RemainingItems, this.Context, func(workItem InputType) (PoolResult, error) {
+		result, runTime, err := this.WorkFunc(workItem)
+		if err != nil {
+			err = fmt.Errorf("Failed to perform individual work on item %v: '%w'.", workItem, err)
+		}
+
+		return PoolResult{workItem, result, runTime, err}, nil
```

Comment:
I'm wondering if we can do more here.

We already take care of caching and locking at the bulk level,
but do you think we can do it here as well?
Like for analysis, we lock for the course, but then further lock for the specific record/analysis.
We could maybe do that here so the caller doesn't have to worry about it.
The caller would need to provide an optional function to produce a lock key from an input type (we already have the retrieval function).
This would also allow us to get pretty accurate timings so the caller wouldn't have to keep track of run time in their work function (mentioned in an earlier comment).